src/terminal/adapter/adaptDispatch.cpp

```diff
@@ -3656,6 +3656,353 @@ void AdaptDispatch::_ReportDECSACESetting() const
+    const auto& cursor = textBuffer.GetCursor();
+    const auto attributes = textBuffer.GetCurrentAttributes();
+
+    // First pull the cursor position relative to the entire buffer out of the console.
+    til::point cursorPosition{ cursor.GetPosition() };
+
+    // Now adjust it for its position in respect to the current viewport top.
+    cursorPosition.y -= viewport.top;
+
+    // NOTE: 1,1 is the top-left corner of the viewport in VT-speak, so add 1.
+    cursorPosition.x++;
+    cursorPosition.y++;
+
+    // If the origin mode is relative, line numbers start at top of the scrolling region.
+    if (_modes.test(Mode::Origin))
+    {
+        cursorPosition.y -= _GetVerticalMargins(viewport, false).first;
+    }
+
+    // Paging is not supported yet (GH#13892).
+    const auto pageNumber = 1;
+
+    // Only some of the rendition attributes are reported.
+    auto renditionAttributes = L'@';
+    renditionAttributes += (attributes.IsIntense() ? 1 : 0);
+    renditionAttributes += (attributes.IsUnderlined() ? 2 : 0);
+    renditionAttributes += (attributes.IsBlinking() ? 4 : 0);
+    renditionAttributes += (attributes.IsReverseVideo() ? 8 : 0);
+    renditionAttributes += (attributes.IsInvisible() ? 16 : 0);
+
+    // There is only one character attribute.
+    const auto characterAttributes = attributes.IsProtected() ? L'A' : L'@';
+
+    // Miscellaneous flags and modes.
+    auto flags = L'@';
+    flags += (_modes.test(Mode::Origin) ? 1 : 0);
+    flags += (_termOutput.IsSingleShiftPending(2) ? 2 : 0);
+    flags += (_termOutput.IsSingleShiftPending(3) ? 4 : 0);
+    flags += (cursor.IsDelayedEOLWrap() ? 8 : 0);
+
+    // Character set designations.
+    const auto leftSetNumber = _termOutput.GetLeftSetNumber();
+    const auto rightSetNumber = _termOutput.GetRightSetNumber();
+    auto charsetSizes = L'@';
+    charsetSizes += (_termOutput.GetCharsetSize(0) == 96 ? 1 : 0);
+    charsetSizes += (_termOutput.GetCharsetSize(1) == 96 ? 2 : 0);
+    charsetSizes += (_termOutput.GetCharsetSize(2) == 96 ? 4 : 0);
+    charsetSizes += (_termOutput.GetCharsetSize(3) == 96 ? 8 : 0);
+    const auto charset0 = _termOutput.GetCharsetId(0);
+    const auto charset1 = _termOutput.GetCharsetId(1);
+    const auto charset2 = _termOutput.GetCharsetId(2);
+    const auto charset3 = _termOutput.GetCharsetId(3);
+
+    using namespace std::string_view_literals;
+
+    // A valid response always starts with DCS 1 $ u.
+    std::wstringstream response;
+    response << L"\033P1$u"sv
+             << cursorPosition.y << ';'
+             << cursorPosition.x << ';'
+             << pageNumber << ';'
+             << renditionAttributes << ';'
+             << characterAttributes << ';'
+             << flags << ';'
+             << leftSetNumber << ';'
+             << rightSetNumber << ';'
+             << charsetSizes << ';'
+             << charset0 << charset1 << charset2 << charset3
+             << L"\033\\"sv; // An ST ends the sequence.
```

Comment:
I would like to avoid using `stringstream`s in this project if possible. They're less performant than `fmt::format`, (somewhat ironically) usually produce more assembly, and are in my personal opinion harder to read than the alternative `fmt` code:
```cpp
const auto response = fmt::format(
    FMT_COMPILE(L"\033P1$u{};{};{};{};{};{};{};{};{};{}{}{}{}\033\\"),
    cursorPosition.y,
    cursorPosition.x,
    pageNumber,
    renditionAttributes,
    characterAttributes,
    flags,
    leftSetNumber,
    rightSetNumber,
    charsetSizes,
    charset0,
    charset1,
    charset2,
    charset3
);
```

I find this easier to read, because I can immediately guestimate how the resulting string will look like. It could argued that it's harder to read, because the association between `{}` and argument is now unclear, but I think that's less worse than the downsides of `stringstream`s. `fmt` also supports named parameter btw like so:
```cpp
using namespace fmt::literals;
const auto response = fmt::format(
    FMT_COMPILE(L"{x}"),
    "x"_a=cursorPosition.x
);
```

But I personally feel like the first snippet is already fine as is, because it's not too difficult to correlate the arguments with the `{}`.

---

This would of course require the
```cpp
inline std::wostream& operator<<(std::wostream& out, const VTID& id)
```
to be changed. An option would be to change `VTID` to be a `union` as mentioned above, so that we can immediately return a `const char*`, which `fmt` will be able to handle. Another option is to write a custom `fmt` formatter like so:
```cpp
template<typename Char>
struct fmt::formatter<VTID, Char>
{
    template<typename FormatContext>
    auto format(const VTID id, FormatContext& ctx) -> decltype(ctx.out())
    {
        uint64_t value = id;
        auto out = ctx.out();

        do
        {
            *out++ = gsl::narrow_cast<Char>(value & 0xFF);
            value >>= CHAR_BIT;
        } while (value);

        return out;
    }
};
```

Reply:
I generally don't like the `fmt` stuff, but I'm happy to change this. I have to admit this your `fmt:format` approach is quite a bit neater than the `stringstream`.